integrating RingCT into sn_dbc.

dan-da opened the pull request:
The primary purpose and functionality of this PR is RingCt.  However, there are many other (mostly) related changes and improvements.

Highlights:

* multisig no longer supported because blst_ringct does not (yet?) support it.  Dbcs are owned by a single SecretKey
* Rng is passed to all random-using fns.  So code can be 100% deterministic.
* issuing genesis Dbc requires that mintnodes use the same deterministic seed.
* AmountSecrets is now a wrapper for blst_ringct::RevealedCommitment
* A mock spentbook is included which generates SpentProofShares and also can generate a random list of DecoyInput
* The MintNodes now must trust pubkey of all SpentBook (sections) in order to validate tx SpentProofs.
* Dbc is now a clent-side only data structure.
* Dbc can optionally be bearer and include the SecretKey inside them.
* Dbc have a base Owner and a derived one-time-use Owner.  the base owner is for interacting with third parties.  The one-time owner is for interacting with mint and spentbook.  The spentbook enforces that a one-time-owner pubkey may only be spent once.
* The spentbook validates Tx of incoming spends.  This serves to keep entries in the SpentBook valid/clean, and also it means that the SpentBook section(s) and Mint reissuing section are double-checking eachother.
* serde support is now behind feature flag "serdes"
* mint-repl now has a nifty reissue_auto command that can generate lots of dbcs randomly.
* There is now a GenesisBuilderMock for easily constructing mintnodes, spentbooks, and the genesis dbc.  Very useful for tests, examples, benches.

dan-da committed:
feat: working on ringct integration. does not build

dan-da committed:
feat: integrate ringct into sn_dbc. wip: it now builds without warnings

dan-da committed:
feat: use PublicKey from KeyManager for genesis dbc owner

dan-da committed:
chore: apply cargo fmt lints

dan-da committed:
fix: use deterministic secret key for genesis dbc

dan-da committed:
feat: re-enable two tests: tests::hash, mint::tests::prop_genesis

dan-da committed:
chore: refactor AmountSecrets to wrap RevealedCommitment. Enable DbcHelper

dan-da committed:
chore: modify TransactionBuilder to wrap RingCtMaterial

dan-da committed:
fix: make ReissueRequestBuilder and DbcBuilder build again

dan-da committed:
chore: remove commented ReissueTransaction code

dan-da committed:
fix: first pass at public_commitments. prop_splitting_the_genesis_dbc() test (mostly) passes

dan-da committed:
fix: handle empty output_amounts in test prop_splitting_the_genesis_dbc

dan-da committed:
feat: include AmountSecrets ciphertext in DbcContent

dan-da committed:
chore: half of prop_dbc_transaction_many_to_many() builds

dan-da committed:
fix: prop_dbc_transaction_many_to_many() is passing now

dan-da committed:
chore: cleanup commented and dead code

dan-da committed:
chore: cargo fmt and clippy

dan-da committed:
feat: return RingCtMaterial from TransactionBuilder::build()

dan-da committed:
chore: impl decoy_inputs for many_to_many test

dan-da committed:
chore: remove commented code

dan-da committed:
chore: use maidsafe blst_ringct

dan-da committed:
fix: make dbc_packet.rs build again

dan-da committed:
fix: refactor to validate spent_proofs in Dbc::confirm_valid

dan-da committed:
fix(reissue): key mint sigs by mlsag index not KeyImage

A big-ish commit.  highlights:

1. KeyImage are not unique across inputs, so we cannot use it as map key.

2. got tests in dbc.rs building and working with ringct

3. mods in builder.rs, mint.rs and dbc.rs to support (1) and (2).

4. fix in builder.rs when checking if SpentProofShares match

5. removed some error variants that are no longer used.

6. added BlsHelper::blstrs_to_blsttc_sk()

7. uncommented NonZeroTinyInt::coerce, which is used by dbc.rs test

dan-da committed:
fix(mint): verify KeyImage unique across inputs

1. Mint reissue now verifies that KeyImage is unique across inputs.

2. Mint reissue now verifies that PublicKey is unique across outputs.

3. reverts KeyImage map changes from previous commit.

4. modifies tests to use unique keys per output/input

dan-da committed:
refactor(dbc): move AmountSecrets to DbcPacket

1. moved AmountSecrets out of Dbc to make it smaller
2. Split DerivedKeySet, avoids Option<SecretKeySet>
3. Added enum DerivedOwner
4. modified tests as necessary

dan-da committed:
feat: add mock spentbook to tests

* add mock spentbook to test module
* merge DbcPacket functionality into Dbc (as both are client-side)
* introduce DerivedOwner for deriving key from base (well known) key
* change type of DbcContent::owner from G1Affine to DerivedOwner
* add owner_derivation_cipher to DbcContent
* modify TransactionBuilder to pair DerivedOwner with each output
* modify DbcBuilder to pair DerivedOwner with each output
* fix: include spent proofs in Dbc::hash()
* improve validation in Dbc::confirm_valid()
* add method to verify secret amount in Dbc matches output commitment
* add SpentProof::to_bytes()
* add (temporary) SpentProof::validate_unsafe()

dan-da committed:
chore: remove unused variants from OwnerBase

DerivedOwner only really needs variants for PublicKey and SecretKey.
The SecretKeySet and PublicKeySet variants made code more verbose
without adding any real value, so they are now removed.

dan-da committed:
chore: remove genesis key from SimpleKeyManager

* Genesis pubkey was unused in SimpleKeyManager, so remove it.
* modify tests accordingly and minor cleanup

dan-da committed:
feat: add spentbook pubkey to mint's key_manager

* add MintNode::trust_spentbook_public_key() to make mint key_manager
  aware of spentbook pubkey
* remove spentbook verifier param from Dbc::confirm_valid()
* add to trait KeyManager fn add_known_key()
* make MintNode::reissue() accept &self instead of &mut self
* make MintNode::issue_genesis_dbc() return self (builder pattern)
* update tests to use MintNode builder pattern and trust spentbook
  pubkey.
* add input_key_image to GenesisDbcShare
* remove genesis_dbc_input static fn
* remove (ugly/temporary) SpentProof::validate_unsafe()

dan-da committed:
refactor: consider first logged tx to be genesis

* makes genesis KeyImage optional in SpentBookMock.
* if genesis key_image not present, first logged tx becomes genesis
* modifies test cases to be a bit cleaner using this facility

dan-da committed:
refactor: consolidate common validation logic

* move common transaction validation logic into new TransactionValidator
* change SimpleKeyManager::new() to impl From
* fix: add genesis public commitments to SpentBookMock with ::set_genesis()
* cleanup and add/edit comments
* cleanup: remove unused Error variants
* cleanup: remove redundant vars from GenesisDbcShare

dan-da committed:
feat: validate tx in spentbook

* impl From<(Amount, Scalar)> for AmountSecrets
* impl Into<Amount> for AmountSecrets
* move genesis creation boilerplate into tests::init_genesis()
* add Tx verification to SpentBookMock::log_spent()
* add SpentBookMock::log_spent_and_skip_tx_verification()
* modify tests to check errors from SpentBookMock::log_spent()
* uncomment test_inputs_are_validated() and gets it working
* uncomment and refactor test_mismatched_amount_and_commitment

dan-da committed:
refactor: rename OwnerBase -> Owner and simplify

* rename OwnerBase -> Owner
* simplify Owner API. Now it derives itself.
* update DerivedOwner accordingly
* rename DbcContent::owner to owner_base
* add detailed doc comments for DbcContent fields

dan-da committed:
feat: add serde feature flag

* reenable serde, now as feature flag 'serdes'
* consolidate blst types into blst.rs and make aliases
* make KeyImage a newtype around G1Affine, usable in a BTreeMap
* obtain AMT_SIZE and BF_SIZE constants from std::mem::size_of()
* make Dbc::owner() public api return blsttc::PublicKey instead of a G1Affine
* add private Dbc::owner_blst() convenience fn.
* remove unused DbcHelper

dan-da committed:
feat: add Dbc::key_image() for checking if spent

Some changes to support mint-repl and general cleanup:

* add KeyImage::to_bytes()
* normalize key fn names to *_secret_key() and *_public_key()
* add Dbc::key_image() fn, needed to check if spent
* add Dbc::amount_secrets() convenience fn
* add Dbc key getter convenience fns.
* remove DbcContent::derive_owner()
* add DbcContent::derivation_index()
* add fns to get blst format keys from DerivedOwner
* fix wrong assertion in SpentBookMock::log_spent()
* simplify tests a bit using new key convenience fns.

dan-da committed:
feat: update mint-repl to use ringct

mint-repl changes:

* enable mint-repl example in Cargo.toml
* rename SpentBook SpentBookNode to match MintNode
* remove DbcUnblinded
* impl reissue_auto command to randomly generate many Dbcs
* remove KeyManager alias for SimpleKeyManager
* remove Signer alias for SimpleSigner
* use RingCtTransaction in place of ReissueTransaction
* disable commands related to multisig usage.
* rework how mint and spentbook nodes are created
* use single char inputs for many user prompts
* support for using SecretKey instead of SecretKeySet
* add newkeys command to generate N new keypairs
* support for ownerless/bearer Dbcs.
* use random decoys from spentbook
* use copy of spentbook code from lib.rs (tests)

dan-da committed:
refactor: merge duplicate spentbook impls

spentbook changes:

* move SpentBookMock from lib into spentbook.rs
* rename SpentBookMock to SpentBookNodeMock
* move SpentBookNodeMock out of tests module
* modify mint-repl to use SpentBookNodeMock
* move random_decoys from mint-repl spentbook to SpentBookNodeMock
* remove mint-repl/spentbook.rs

dan-da committed:
refactor: improve Dbc ownership API

* update mint-repl to use Owner changes
* update bulders to use Owner changes
* rename DerivedOwner to OwnerOnce
* add doc comments for Dbc describing ownership
* remove redundant key getters in Dbc
* adder bearer methods in Dbc
* remove redundant key getters in OwnerOnce
* renamed derived_owner.rs to owner.rs
* add blst getters to Owner
* add basic doc comments to Owner, OwnerOnce
* remove redundant secret_key field from GenesisDbcShare

dan-da committed:
improvement: make spentbook more efficient

change the way data is stored and queried in SpentBookNodeMock

* store only a single copy of each Tx. tx_hash    -> tx
* store key_image index.               key_image  -> tx_hash
* store public_key index:              public_key -> output_proof
* avoid full table scan when matching input pks with output pk.
* add some explanation in doc comment

dan-da committed:
docs: add comments around trusting spentbook keys

dan-da committed:
feat: add GenesisBuilderMock

This commit adds a builder for initializing a set of N mintnodes, a set
of Y spentbooks, and generating a genesis dbc with amount Z.

This functionality is need in tests, mint-repl, and benches. As such,
it cannot be in tests, but must be part of sn_dbc library.  It is
planned that a future commit will refactor all the "mock" items into
a mock module, behind a feature flag.

* impl builder::mock::GenesisBuilderMock
* use GenesisBuilderMock in mint-repl mk_new-mint()
* use GenesisBuilderMock in lib::tests::issue_genesis()
* improve how rng parameters are passed

dan-da committed:
test: get reissue bench working with ringct

* re-enable reissue bench in Cargo.toml
* use latest criterion 0.3.5
* update benches/reissue.rs to use ringct and be deterministic
* impl ReissueRequestBuilder::add_spent_proof_shares()

dan-da committed:
docs: update/add copyright notice in source files

dan-da committed:
chore: remove init_genesis() in lib.rs

rationale: now that GenesisBuilderMock exists, init_genesis() in lib.rs
was just a wrapper function for compatibility with existing tests.  So
this is a minor cleanup to make the tests use GenesisBuilderMock
directly.

* remove lib.rs init_genesis() wrapper fn
* remove unused allow clippy::from_iter_instead_of_collect
* use GenesisBuilderMock::init_genesis_single in tests

dan-da committed:
style: disambiguate KeyImage usage

The KeyImage wrapper for G1Affine was being used in a few places
where a PublicKeyBlst is used as map key.  But it is semantically
wrong to call it a key image.

This commit changes the wrapper name to PublicKeyBlstMappable and
makes KeyImage a type alias of it.  So now the appropriate name
is used according to context.

dan-da committed:
docs: add some doc comments

dan-da committed:
style: add input Dbc(s) in TransactionBuilder

* add some Doc comments to TransactionBuilder
* add Dbc::as_true_input()
* add TransactionBuilder fns to add Dbc input

dan-da committed:
test: use add_input_dbc*() in mint-repl and tests

* remove AsRef in add_input_dbc* fn signatures
* use add_input_dbc*() in mint-repl and tests
* add some missing doc comments

dan-da committed:
fix: avoid possible panic in issue_genesis_dbc

Various cleanup:

* add some missing doc comments
* return error instead of panicing in issue_genesis_dbc
* use (3) random_decoys() in tests.
* remove commented code from builder.rs
* make panic in SpentBookMock::iter() explicit

dan-da committed:
test: revert change to num_outputs in benchmark

The bench originally tested with 100 outputs, but was set to 2
in a recent commit.  This reverts the value to 100.

davidrusu commented on src/dbc.rs:
```diff
@@ -1,96 +1,309 @@
             Err(Error::DbcContentNotPresentInTransactionOutput)
         } else {
             Ok(())
         }
```
Should we also call `self.confirm_provided_amount_matches_commitment(..)`?

dan-da replied:
maybe, but that would be "invalid" in a different sense than other confirm_valid() errors.

Presently we can distinguish between:
1. Dbc is valid and amount secrets are valid
2. Dbc is valid and amount secrets are invalid
3. Dbc is invalid and amount secrete are invalid
4. Dbc is invalid and amount secrets are valid

Of these, I think 1 and 2 are most important.  Not much need to distinguish between 3 and 4 usually I guess.

One way to retain same granularity would be to return an Enum from confirm_valid() instead of just ().

Or maybe we just keep the present API.

Anyway, I will open an issue for this.

dan-da replied:
created #127

davidrusu commented on src/dbc.rs:
```diff
@@ -186,142 +403,120 @@ mod tests {
-                (genesis_key, genesis_sig),
-            )]),
-        };
+        let input_owners: Vec<OwnerOnce> = (0..=n_inputs.coerce())
```
Did you mean n_inputs + 1 owners?
```suggestion
        let input_owners: Vec<OwnerOnce> = (0..n_inputs.coerce())
```

dan-da replied:
yeah, that works.

davidrusu commented on src/dbc.rs:
```diff
@@ -186,142 +403,120 @@ mod tests {
-                (genesis_key, genesis_sig),
-            )]),
-        };
+        let input_owners: Vec<OwnerOnce> = (0..=n_inputs.coerce())
+            .map(|_| OwnerOnce::from_owner_base(Owner::from_random_secret_key(&mut rng), &mut rng8))
```
Also, I've been liking `std::iter::repeat_with(|| ...).take(n)` as a replacement for `(0..n).into_iter().map(|_| ...)`.

dan-da replied:
interesting.  I think I will leave it as-is for now though.   it's actually 1 extra line to do it the repeat_with() way, after cargo fmt.

```
        let input_owners: Vec<OwnerOnce> = std::iter::repeat_with(|| {
            OwnerOnce::from_owner_base(Owner::from_random_secret_key(&mut rng), &mut rng8)
        })
        .take(n_inputs.coerce())
        .collect();
```
vs.
```
        let input_owners: Vec<OwnerOnce> = (0..n_inputs.coerce())
            .map(|_| OwnerOnce::from_owner_base(Owner::from_random_secret_key(&mut rng), &mut rng8))
            .collect();
```

davidrusu commented on src/dbc.rs:
```diff
@@ -186,142 +403,120 @@ mod tests {
+        for (i, (mlsag, _mlsag_material)) in reissue_tx
+            .mlsags
+            .iter()
+            .zip(material.inputs.iter())
```
Why zip if we are not using the `mlsag_material`?

dan-da replied:
fixed.

davidrusu commented on src/dbc.rs:
```diff
@@ -186,142 +403,120 @@ mod tests {
             reissue_share
                 .mint_node_signatures
-                .keys()
+                .into_iter()
```
Seems like we don't need the values, can we keep `.keys()`?

dan-da replied:
the into_iter() avoided the need to clone() each in_owner inside the map().

I've changed it to use into_keys() instead.

davidrusu commented on src/dbc.rs:
```diff
@@ -330,89 +525,117 @@ mod tests {
-            fuzzed_transaction_sigs
-                .insert(rand::random::<SpendKey>(), (genesis_key, mint_sig.clone()));
+            fuzzed_transaction_sigs.insert(
+                Default::default(),
```
Feels a bit loose to have a `Default` impl on `KeyImage`, How about we just be explicit with a `KeyImage::random(&mut rng)` and drop the Default?

dan-da replied:
done.   added ::random(), and removed derive Default.

```
    pub fn random(rng: &mut impl rand8::RngCore) -> Self {
        Self(G1Projective::random(rng).to_affine())
    }
```

davidrusu commented on src/dbc.rs:
```diff
@@ -330,89 +525,117 @@ mod tests {
+            .public_key_blst();
+
+        // Check if commitment in AmountSecrets matches the commitment in tx OutputProof
+        let commitments_match = dbc
+            .confirm_provided_amount_matches_commitment(
+                &owner_once.owner_base().secret_key()?,
+                &fuzzed_amt_secrets,
+            )
+            .is_ok();
```
If we move this confirmation into `confirm_valid()`, we should remember to remove this

dan-da replied:
commented in #127

davidrusu commented on src/spent_proof.rs:
```diff
@@ -1,56 +1,57 @@
 
     pub spentbook_sig_share: NodeSignature,
+
+    pub public_commitments: Vec<Commitment>,
```
We need to be including the public_commitments in the spent proof signature. Otherwise mint's can't trust that the public commitments here were checked for validity by the spent book nodes

dan-da replied:
ok, this is a bit more involved and I think requires mods in spentbook.rs and also validation.rs.    leaving for tmw.

dan-da replied:
ok, I made a SpentProofContent to hold all the data that SpentBook needs to sign, and it now holds the transaction hash.
```
pub struct SpentProofContent {
    pub key_image: KeyImage,
    pub transaction_hash: Hash,
    pub public_commitments: Vec<Commitment>,
}
```
So `SpentProofShare` and `SpentProof` both hold a `SpentProofContent` plus signing info.

confirm_valid() also verifies that the incoming tx hash matches this one.  And did some cleanup/refactoring in spent_proof.rs.  So I think it's good now.

commit 59d7f137bb5275b50dd7c6c4aedd06ee5cc85702

davidrusu commented on src/validation.rs:
```diff
@@ -0,0 +1,147 @@
+        spent_proofs: &BTreeSet<SpentProof>,
+    ) -> Result<(), Error> {
+        if transaction.mlsags.is_empty() {
+            return Err(Error::TransactionMustHaveAnInput);
```
This check should be moved to `RingCtTransaction::verify()` if it's not already there.

dan-da replied:
created #128

davidrusu commented on src/validation.rs:
```diff
@@ -0,0 +1,147 @@
+            return Err(Error::SpentProofInputMismatch);
+        }
+
+        // Verify that each KeyImage is unique in this tx.
+        let keyimage_unique: BTreeSet<KeyImage> = transaction
+            .mlsags
+            .iter()
+            .map(|m| m.key_image.into())
+            .collect();
+        if keyimage_unique.len() != transaction.mlsags.len() {
+            return Err(Error::KeyImageNotUniqueAcrossInputs);
+        }
```
This check should be moved to `RingCtTransaction::verify()` if it's not already there.

dan-da replied:
created #126

dan-da replied:
added comment in #128

davidrusu commented on src/validation.rs:
```diff
@@ -0,0 +1,147 @@
+        }
+
+        // Verify that each pubkey is unique in this transaction.
+        let pubkey_unique: BTreeSet<KeyImage> = transaction
+            .outputs
+            .iter()
+            .map(|o| (*o.public_key()).into())
+            .collect();
+        if pubkey_unique.len() != transaction.outputs.len() {
+            return Err(Error::PublicKeyNotUniqueAcrossOutputs);
+        }
```
This check should be moved to `RingCtTransaction::verify()` if it's not already there.

dan-da replied:
created #130

davidrusu reviewed (commented):
Left a few comments that we need to fix, but it's looking good.

dan-da committed:
chore: make sn_dbc::Error (de)serializable

dan-da commented on examples/mint-repl/mint-repl.rs:
```diff
@@ -878,62 +773,160 @@ fn prepare_reissue_cli(mintinfo: &mut MintInfo) -> Result<()> {
+        readline_prompt_nl_default("\nHow many reissues to perform [10]: ", "10")?.parse()?;
+
+    for _i in 1..=num_reissues {
+        let max_inputs = vec![mintinfo.reissue_auto.unspent_dbcs.len(), 10]
+            .into_iter()
+            .min()
+            .unwrap_or(10);
```
done.

dan-da commented on examples/mint-repl/mint-repl.rs:
```diff
@@ -878,62 +773,160 @@ fn prepare_reissue_cli(mintinfo: &mut MintInfo) -> Result<()> {
+
+            tx_builder = tx_builder.add_input_dbc(dbc, &base_sk, decoy_inputs, &mut rng8)?;
+
+            mintinfo.reissue_auto.unspent_dbcs.remove(&dbc.hash());
```
done.

dan-da commented on examples/mint-repl/mint-repl.rs:
```diff
@@ -878,62 +773,160 @@ fn prepare_reissue_cli(mintinfo: &mut MintInfo) -> Result<()> {
+        while tx_builder.outputs_amount_sum() < inputs_sum {
+            // randomize output amount
+            let diff = inputs_sum - tx_builder.outputs_amount_sum();
+            let amount = if diff == 1 { 1 } else { rng.gen_range(1, diff) };
```
done.  though I had to modify the while loop condition like so:
```
while tx_builder.outputs_amount_sum() < inputs_sum || tx_builder.outputs().is_empty() {
```
otherwise zero valued outputs can become inputs which sum to zero, so the while body does not exec, leading to an invalid tx with zero outputs.

that's a very subtle one.

dan-da commented on examples/mint-repl/mint-repl.rs:
```diff
@@ -878,62 +773,160 @@ fn prepare_reissue_cli(mintinfo: &mut MintInfo) -> Result<()> {
+            );
+        }
+
+        let (reissue_tx, revealed_commitments, _material, output_owners) =
+            tx_builder.build(&mut rng8)?;
```
created #122 for this.

dan-da commented on examples/mint-repl/mint-repl.rs:
```diff
@@ -878,62 +773,160 @@ fn prepare_reissue_cli(mintinfo: &mut MintInfo) -> Result<()> {
-    output_pk_pks: &HashMap<PublicKey, PublicKeySet>,
-) -> Result<()> {
-    let mut dbc_builder = DbcBuilder::new(reissue_request.transaction.clone());
+fn reissue(mintinfo: &mut MintInfo, reissue_request: ReissueRequestUnblinded) -> Result<()> {
```
done.   /s/Unblinded/Revealed/g

dan-da commented on src/amount_secrets.rs:
```diff
@@ -0,0 +1,203 @@
+    }
+
+    /// build AmountSecrets from fixed size byte array.
+    pub fn from_bytes(bytes: [u8; AMT_SIZE + BF_SIZE]) -> Self {
```
created #123 for this.

dan-da commented on src/amount_secrets.rs:
```diff
@@ -0,0 +1,203 @@
+    }
+
+    /// build AmountSecrets from byte array reference
+    pub fn from_bytes_ref(bytes: &[u8]) -> Result<Self, Error> {
```
not really.  I forget the details, but there was some ugliness that made me think its better just to keep to_bytes and from_bytes as regular impl methods.

we may want to revisit later, to make standard to/from bytes and also hash() across structs and crates.

dan-da commented on src/blst.rs:
```diff
@@ -0,0 +1,131 @@
+
+impl Eq for PublicKeyBlstMappable {}
+
+impl PartialOrd for KeyImage {
```
gotta keep clippy happy.

dan-da commented on src/blst.rs:
```diff
@@ -0,0 +1,131 @@
+    }
+}
+
+impl Hash for PublicKeyBlstMappable {
```
yep, you got it.

dan-da commented on src/builder.rs:
```diff
@@ -1,144 +1,275 @@
-    pub reissue_transaction: ReissueTransaction,
-    pub spent_proof_shares: BTreeMap<SpendKey, HashSet<SpentProofShare>>,
+    pub transaction: RingCtTransaction,
+    pub spent_proof_shares: BTreeMap<usize, HashSet<SpentProofShare>>,
```
This is done because the commitments must be in original input order for this code in TransactionValidator to work:
```
        let public_commitments: Vec<Vec<Commitment>> = spent_proofs_sorted
            .iter()
            .map(|s| s.public_commitments.clone())
            .collect();

        transaction.verify(&public_commitments)?;
```
Looking at it again, I should be able to get them in proper order by matching with the tx.mlsags inputs, which are ordered.   anyway, it's a big enough change I will do it in a separate commit, 2nd pass.

dan-da replied:
done.  in 08e5883755c5707d4be626f494bcd60d9e89abfa

dan-da committed:
feat: add input/output getters to Tx builder

This enables api users (such as mint-repl) to check the input/output
len (or inspect ins/outs) while building up the Tx.

dan-da committed:
chore: impl reviewer suggestions from pr 121

* add assert
* change how reissue_auto ranomizes inputs and outputs
* allows amount==0 outputs in reissue_auto
* delay unspent removal in reissue_auto until after tx is reissued.
* change *Unblinded to *Revealed for consistency with RingCt usage

dan-da commented on src/builder.rs:
```diff
@@ -1,144 +1,275 @@
 
-                if shares
+                if !shares
                     .iter()
                     .map(SpentProofShare::spentbook_pks)
-                    .any(|pks| pks != any_share.spentbook_pks())
+                    .any(|pks| pks == any_share.spentbook_pks())
```
done. nice catch.

dan-da commented on src/builder.rs:
```diff
@@ -1,144 +1,275 @@
                 }
 
-                let spent_sig = any_share.spent_sig.clone();
+                if !shares
+                    .iter()
+                    .map(|s| &s.public_commitments)
+                    .any(|pc| *pc == any_share.public_commitments)
+                {
+                    return Err(Error::ReissueRequestPublicCommitmentMismatch);
+                }
```
done. nice catch.

re future work:  perhaps.  hopefully we can make it possible for client to call RingCtTransaction::verify() before sending tx to spentbook or mint.  If we can do that, maybe we don't need any other build validations...

dan-da replied:
opened #124

dan-da committed:
fix: reverse logic valdiating spentproof shares

caught by reviewer of pr 121.

The previous code was only checking that one share matches, not all
shares

dan-da commented on src/builder.rs:
```diff
@@ -224,19 +365,18 @@ impl DbcBuilder {
-            for input in self.reissue_transaction.inputs.iter() {
-                if rs.mint_node_signatures.get(&input.spend_key()).is_none() {
+            for mlsag in rs.transaction.mlsags.iter() {
+                if !rs
+                    .mint_node_signatures
+                    .keys()
+                    .any(|k| *k == mlsag.key_image.into())
```
done.

dan-da commented on src/builder.rs:
```diff
@@ -258,37 +398,349 @@ impl DbcBuilder {
+                let mut rng8 = rng8_initial.clone();
+
+                let (mut mint_node, genesis_dbc_share) =
+                    mint_node.issue_genesis_dbc(self.genesis_amount, &mut rng8)?;
```
opened #125

dan-da committed:
style: express validation logic differently

See https://github.com/maidsafe/sn_dbc/pull/121#discussion_r805181871

dan-da commented on src/dbc.rs:
```diff
@@ -1,96 +1,309 @@
+/// index is generated and used to derive a one-time-use Owner Once.
+///
+/// The Owner Once is used for a single transaction only and must be unique
+/// within a transaction.
```
changed to:
> within the transaction as well as globally for the output DBC's to be spendable.

dan-da committed:
style: rename transaction_sigs to mint_sigs

See https://github.com/maidsafe/sn_dbc/pull/121#discussion_r805182798

dan-da commented on src/dbc.rs:
```diff
@@ -1,96 +1,309 @@
-    pub transaction: DbcTransaction,
-    pub transaction_sigs: BTreeMap<SpendKey, (PublicKey, Signature)>,
+    pub transaction: RingCtTransaction,
+    pub transaction_sigs: BTreeMap<KeyImage, (PublicKey, Signature)>,
```
done.  more diffs, yay!

dan-da commented on src/dbc.rs:
```diff
@@ -1,96 +1,309 @@
+
+    /// returns KeyImage for the owner's derived public key
+    /// This is useful for checking if a Dbc has been spent.
+    pub fn key_image(&self, base_sk: &SecretKey) -> Result<KeyImage> {
```
created #126

dan-da committed:
chore: cleanup PublicKeyBlstMappable interface

implementing reviewer suggestions from PR 121

* remove derive Default from PublicKeyBlstMappable
* add PublicKeyBlstMappable::random()
* avoid unneeded Owner key in test case
* eliminate unnecessary .zip() in test case
* use into_keys instead of into_iter in test case
* use KeyImage::random instead of ::default() in test case

dan-da committed:
fix: refactor to include tx_hash in spentbook sig

* add SpentProofContent to represent all data to be signed by SpentBook
* SpentProof[Share] now includes transaction_hash in spentbook's
  signature
* refactor SpentProofShare and SpentProof to use SpentProofContent
* add doc comments for SpentProof{*]
* add Error variant InvalidTransactionHash for case when input tx does
  not match tx signed by spentbook
* change ReissueRequestBuilder::spent_proof_shares map value from HashSet
  to BTreeSet so we do not need to impl Hash for SpentProofShare

dan-da committed:
style: key by KeyImage instead of input index

suggested by pr reviewer:
  https://github.com/maidsafe/sn_dbc/pull/121#discussion_r806243966

* key SpentProofShare by KeyImage instead of input idx in
  ReissueRequestBuilder
* adapt calling code as necessary

davidrusu reviewed (approved):
Looks good!
